### DISCUSSION TITLE
Support for alternate keyboard layouts

### DISCUSSION THREAD

**Discussion by gerardroche** [NeoVintageous/NeoVintageous#926] [General]
- [x]  Support Colemak mappings #918 
- [x]  Support Dvorak mappings #919 

I'm thinking we could autoload the mappings from any of the installed packages above **before** the user rc file is loaded. This allows the user to install an existing layout like Colemak or Dvorak, but have the ability override any of those mappings via the `*unmap` commands or remap any of the keys themselves.

This would be useful because, for example, you might like most of mappings from the existing package, but might like unset or remap one or two.

So... I'm thinking either

1. The user installs the package and then manually tells neovintageous to use it via a setting

    The downside to this is that the user has to configure neovintageous to load the mappings.

    ```
    "neovintageous.use_layout": "dvorak"
    ``` 
    
    or path to file

    ```
    "neovintageous.use_layout": "Packages/Dvorak/dvorak.neovintageous"
    ```

    or allow both. A values starting `Packages/` would be assumed to be a path. Otherwise `"neovintageous.use_layout": "{name}" would look for:

    ```
    "neovintageous.use_layout": "Packages/{name}/{lowercase_name}.neovintageous"
    ```

2. Auto discovery

    a) auto discover by looking for `*.neovintageous` files and loading them ordered by name.
    b) the packages themselves set a setting e.g. `neovintageous.dvorak.enable`. We would only be able to auto discover known pakages this way. NeoVintageous would specifically look for dvorak or colemak or any known package and if enabled i.e. if it's installed, load its `*.neovintageous` file.

   - We could allow auto discovery to be disabled.
   - And allow functionality of 1 above too. If a layout is specifically set then it would disable auto discovery.

I not sure I like 2a because it would allow any plugin to set mappings, which sounds good but I don't know.

I think 2b is nice because it means a user can install any of the packages above and they will automatically work without having to configure anything.

I think 1 is flexible to allow users to make their own packages or  even use a third party one unknown to neovintageous. 

A mix of 1 and 2b maybe?

**Comment by sourcevault** [comment]
Keep it simple and use 1 b.

```
"neovintageous.use_layout": "Packages/Dvorak/dvorak.neovintageous"
```

If you need to, add all the extra stuff after having core functionality.

![layout_ergodox](https://github.com/NeoVintageous/NeoVintageous/assets/20347647/eabcec72-e377-4087-865d-e47f03eb7bef)

☝️ btw I am using a crazy layout 🤠, its much better than either, since `Esc` is in home row, super comfee when using `NeoVintageous`.

I would not be surprised at all if a large number of Colemak,  Dvorak users eventually switch to a much more personalized keyboard. 

![giphy](https://github.com/NeoVintageous/NeoVintageous/assets/20347647/c9ea0302-4455-4bb9-91c5-d494e1b04cb4)

Warm Regards.
